assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js

```diff
@@ -0,0 +1,147 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import { useEffect, useRef, useState } from 'react';
+import { useFeature } from 'flagged';
+
+/**
+ * Internal dependencies
+ */
+import { useAPI } from '../../api';
+import { useConfig } from '../../config';
+
+/**
+ * Function returning the current Meta Boxes DOM Node in the editor
+ * whether the meta box area is opened or not.
+ * If the MetaBox Area is visible returns it, and returns the original container instead.
+ *
+ * @param {string} location Meta Box location.
+ * @return {Element} Container
+ */
+function getMetaBoxContainer(location) {
+  // Class name as set in <MetaBoxesArea>
+  const area = document.querySelector(
+    `.web-stories-meta-boxes-area-${location} .metabox-location-${location}`
+  );
+
+  return (
+    area || document.querySelector('#metaboxes .metabox-location-' + location)
+  );
+}
+
+// Returns a prop's previous value.
+function usePrevious(value) {
+  const ref = useRef();
+  useEffect(() => {
+    ref.current = value;
+  });
+  return ref.current;
+}
+
+/**
+ * Effect to save meta boxes for a story.
+ *
+ * @see https://github.com/WordPress/gutenberg/blob/148e2b28d4cdd4465c4fe68d97fcee154a6b209a/packages/edit-post/src/store/effects.js#L24-L126
+ *
+ * @param {Object} props Hook props.
+ * @param {Object} props.story Story object.
+ * @param {boolean} props.isSaving Whether saving is in progress.
+ * @param {boolean} props.isAutoSaving Whether autosaving is in progress.
+ * @return {{isSavingMetaBoxes: boolean}} Metaboxes status.
+ */
+function useSaveMetaBoxes({ story, isSaving, isAutoSaving }) {
+  const isFeatureEnabled = useFeature('customMetaBoxes');
+  const { metaBoxes = {} } = useConfig();
+
+  const locations = ['normal', 'advanced'];
+  const hasMetaBoxes = locations.some((location) =>
+    Boolean(metaBoxes[location]?.length)
+  );
+
+  const {
+    actions: { saveMetaBoxes },
+  } = useAPI();
+
+  const [isSavingMetaBoxes, setIsSavingMetaBoxes] = useState(false);
+
+  const wasSaving = usePrevious(isSaving);
+  const wasAutoSaving = usePrevious(isAutoSaving);
+
+  // Save metaboxes when performing a full save on the post.
+  useEffect(() => {
+    if (
+      !hasMetaBoxes ||
+      !isFeatureEnabled ||
+      isSaving ||
+      isAutoSaving ||
+      isSavingMetaBoxes ||
+      !wasSaving ||
+      wasAutoSaving
+    ) {
+      return;
+    }
+
+    async function save() {
+      // Saves the wp_editor fields.
+      global.tinyMCE?.triggerSave();
+
+      // We gather all the metaboxes locations data and the base form data.
+      const baseFormElement = document.querySelector('.metabox-base-form');
+      const baseFormData = new global.FormData(baseFormElement || undefined);
+
+      const formDataToMerge = [
+        baseFormData,
+        ...['normal', 'advanced'].map(
```

Comment:
Isn't this a repeat of ` const locations `?

Reply:
Yep, good catch!

Reply:
⬆️